### PR TITLE
fix: Special device migration

### PR DESCRIPTION
**Pull request by aterga**
# Motivation

This PR adds special case handling to the anchor device migration. 

**Context**

With the previous migration code in `impl From<Anchor> for (StorableFixedAnchor, StorableAnchor)`, all anchors were migrated to unbounded memory (which is not currently used). During this process, some inconsistent keys were discovered.
* credential_id = null (60135 cases)
    * key type = BrowserStorageKey (56822 cases)
    * key type = Unknown (3308 cases)
    * key type = Platform (5 cases)

This PR adds a handler for all the special cases above to the migration and repeat the migration in the next release.

Main condition: if there is a credential ID, fall back to passkey; otherwise, treat this as a recovery phrase.

For the special cases, the (credential_id, purpose, key_type) triplets is associated with the migrated data, so there is no risk of data loss.

# Changes

* Introduced `SpecialDeviceMigration` struct to track inconsistent device configurations during migration.
* Added optional `special_device_migration` field to both `StorablePasskeyCredential` and `StorableRecoveryKey` structs.
* Updated migration logic in `From<Anchor>` implementation to handle edge cases explicitly.

# Tests

* Adjusted to cover some inconsistent corner cases (`should_handle_devices_without_credential_id`).

### CODE REVIEW

File: src/internet_identity/src/storage/anchor.rs [hunk 1]
```diff
@@ -176,7 +178,37 @@ impl From<Anchor> for (StorableFixedAnchor, StorableAnchor) {
...
+                        (
+                            is_recovery_phrase,
+                            Some(SpecialDeviceMigration {
+                                credential_id: credential_id.clone().map(ByteBuf::into_vec),
+                                purpose: purpose.clone().into(),
+                                key_type: key_type.clone().into(),
+                            }),
+                        )
+                    }
+                };
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/storage/anchor.rs]: It is less error prone in this case to aggregate all unhappy cases under one arm of code, since we would need to analyze these cases and decide how to classify and potentially treat them post-migration.

[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/storage/anchor.rs]: If there is a credential ID, this key could be used as a passkey. 

This is a safer fallback due to the fact that authentication checks in the backend iterate through all keys, not only those explicitly marked as recovery phrases. This includes passkeys. So even if the migration would mis-classify these cases, the user would still be able to use the keys as recovery phrases in the unlikely event that they added the credential IDs to what in reality actually was a recovery phrase by accident.

File: src/internet_identity/src/storage/storable/special_device_migration.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    Unknown,
+    #[n(1)]
+    Platform,
+    #[n(2)]
+    CrossPlatform,
+    #[n(3)]
+    SeedPhrase,
+    #[n(4)]
+    BrowserStorageKey,
+}
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/storage/storable/special_device_migration.rs]: Obvious

File: src/internet_identity/src/storage/anchor/tests.rs [hunk 3]
```diff
@@ -976,14 +1018,51 @@ mod from_conversion_tests {
...
+        assert_eq!(
+            passkeys[0].special_device_migration,
+            Some(SpecialDeviceMigration {
+                credential_id: None,
+                purpose: Purpose::Authentication.into(),
+                key_type: KeyType::BrowserStorageKey.into(),
+            })
+        );
+        assert_eq!(passkeys[0].credential_id, vec![0xde, 0xad, 0xbe, 0xef]);
+    }
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/storage/anchor/tests.rs:1057]: There's no added benefit of testing the different logging behaviors in the two cases.

File: src/internet_identity/src/main.rs [hunk 4]
```diff
@@ -77,10 +78,22 @@ thread_local! {
...
 }
 
+#[query(hidden = true)]
+fn get_anchor_migration_special_cases(anchor_number: AnchorNumber) -> Vec<SpecialDeviceMigration> {
+    ANCHOR_MIGRATION_SPECIAL_CASES
+        .with_borrow(|cases| cases.get(&anchor_number).cloned().unwrap_or_default())
+}
+
+#[query(hidden = true)]
+fn get_anchor_migration_special_cases_keys() -> Vec<AnchorNumber> {
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity/src/main.rs:93]: This is already public information with the `lookup` method. The `get_anchor_migration_special_cases_keys` method is hidden not for access control reasons, but documentation reasons, it's temporary and will be removed once it has served it's purpose (debugging the data migration).

File: src/internet_identity/src/main.rs [hunk 5]
```diff
@@ -77,10 +78,22 @@ thread_local! {
     pub(crate) static RECOVERY_PHRASE_MIGRATION_BATCH_ID: RefCell<u64> = const { RefCell::new(0) };
     pub(crate) static RECOVERY_PHRASE_MIGRATION_ERRORS: RefCell<Vec<String>> = const { RefCell::new(Vec::new()) };
     pub(crate) static RECOVERY_PHRASE_MIGRATION_LAST_ANCHOR_ID: RefCell<Option<u64>> = const { RefCell::new(None) };
+    pub(crate) static ANCHOR_MIGRATION_SPECIAL_CASES: RefCell<BTreeMap<AnchorNumber, Vec<SpecialDeviceMigration>>> = const { RefCell::new(BTreeMap::new()) };
 
     static TIMER_ID: RefCell<Option<TimerId>> = const { RefCell::new(None) };
 }
 
+#[query(hidden = true)]
+fn get_anchor_migration_special_cases(anchor_number: AnchorNumber) -> Vec<SpecialDeviceMigration> {
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/main.rs:87]: [Ditto](https://github.com/dfinity/internet-identity/pull/3530#discussion_r2606747541)
